syncer/syncer.go

```diff
@@ -1803,35 +1752,22 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext) e
 		// NOTE: if we need singleton Syncer (without dm-master) to support sharding DDL sync
 		// we should add another config item to differ, and do not save DDLInfo, and not wait for ddlExecInfo
 
-		ddlInfo1 := &pb.DDLInfo{
-			Task:   s.cfg.Name,
-			Schema: ddlInfo.tableNames[1][0].Schema, // use target schema / table name
-			Table:  ddlInfo.tableNames[1][0].Name,
-			DDLs:   needHandleDDLs,
+		// construct & send shard DDL info into etcd, DM-master will handle it.
+		shardInfo := s.pessimist.ConstructInfo(ddlInfo.tableNames[1][0].Schema, ddlInfo.tableNames[1][0].Name, needHandleDDLs)
+		rev, err2 := s.pessimist.PutInfo(shardInfo)
+		if err2 != nil {
+			return err2
 		}
-		s.ddlInfoCh <- ddlInfo1 // save DDLInfo, and dm-worker will fetch it
+		shardLockResolving.WithLabelValues(s.cfg.Name).Set(1) // block and wait DDL lock to be synced
+		s.tctx.L().Info("putted shard DDL info", zap.Stringer("info", shardInfo))
 
-		// block and wait DDL lock to be synced
-		shardLockResolving.WithLabelValues(s.cfg.Name).Set(1)
-		for {
-			var ok bool
-			ddlExecItem, ok = <-s.ddlExecInfo.Chan(needHandleDDLs)
-			if !ok {
-				// chan closed
-				shardLockResolving.WithLabelValues(s.cfg.Name).Set(0)
-				s.tctx.L().Warn("canceled from external", zap.String("event", "query"), zap.String("source", source), zap.Strings("ddls", needHandleDDLs), zap.ByteString("raw statement", ev.Query), zap.Stringer("start position", startPos), log.WrapStringerField("end position", ec.currentPos))
-				return nil
-			} else if len(ddlExecItem.req.DDLs) != 0 && !reflect.DeepEqual(ddlExecItem.req.DDLs, needHandleDDLs) {
-				// ignore un-cleared cached/duplicate DDL execute request
-				// check `len(ddlExecItem.req.DDLs) != 0` to support old DM-master and `break-ddl-lock`
-				s.tctx.L().Warn("ignore mismatched DDL execute request", zap.String("source", source), zap.Strings("expect", needHandleDDLs), zap.Strings("request", ddlExecItem.req.DDLs))
-				continue
-			}
-			shardLockResolving.WithLabelValues(s.cfg.Name).Set(0)
-			break
+		shardOp, err2 := s.pessimist.GetOperation(ec.tctx.Ctx, shardInfo, rev)
+		shardLockResolving.WithLabelValues(s.cfg.Name).Set(0)
```

Comment:
do this after err2 is nil?

Reply:
YES, if `err2 == nil`, syncer will abort the replication, then it's not waiting for the lock to be resolved anymore, but can be in resolving after resumed without error.